### PR TITLE
Allow to deserialize more valid RFC3339 date-time and make the format customizable

### LINKED ISSUE #5382: [java-generator] date-time doesn't parse valid RFC 3339 timestamps
### Describe the bug

When autogenerating Java classes from a CRD yaml using maven plugin, the format being added in the JSON parser is [this](https://github.com/fabric8io/kubernetes-client/blob/a7c630bbd368a976d6b0e72287dcf4f7348c54bd/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java#L221). 
This, however, is not sufficient and fails for some of the valid timestamps with timezone offset like `2023-07-07T06:43:19+00:00`, which would need the format to be `yyyy-MM-dd'T'HH:mm:ssXXX`([RFC standard](https://datatracker.ietf.org/doc/html/rfc3339#section-5.6)).
Can we please either correct this to include more formats, or add an option to pass additional formats?

### Fabric8 Kubernetes Client version

6.8.0

### Steps to reproduce

Use the maven plugin to autogenerate classes from a CRD yaml with the following type for field:
type: string
format: date-time

Now try to parse resources with timestamps of format `yyyy-MM-dd'T'HH:mm:ssXXX` with these classes.

The following exception is thrown: 
```
Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.time.ZonedDateTime` from String "2023-07-07T06:43:19+00:00": 
Failed to deserialize java.time.ZonedDateTime: (java.time.format.DateTimeParseException) Text '2023-07-07T06:43:19+00:00' could not be parsed, unparsed text found at index 22
```

### Expected behavior

The timestamp should get parsed successfully into the relevant field.

### Runtime

Kubernetes (vanilla)

### Kubernetes API Server version

1.25.3@latest

### Environment

Linux

### Fabric8 Kubernetes Client Logs

```shell
com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.time.ZonedDateTime` from String "2023-08-07T06:43:19+00:00": Failed to deserialize java.time.ZonedDateTime: (java.time.format.DateTimeParseException) Text '2023-08-07T06:43:19+00:00' could not be parsed, unparsed text found at index 22
[1:18 pm, 11/08/2023]: org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:108)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88)
	Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.time.ZonedDateTime` from String "2023-08-07T06:43:19+00:00": Failed to deserialize java.time.ZonedDateTime: (java.time.format.DateTimeParseException) Text '2023-08-07T06:43:19+00:00' could not be parsed, unparsed text found at index 22
 at [Source: (BufferedInputStream); line: 1, column: 3077] (through reference chain: io.fabric8.kubernetes.api.model.DefaultKubernetesResourceList["items"]->java.util.ArrayList[0]->com.example.v1.MyCRD["status"]-com.example.v1.mycrdstatus.Pod["last_updated"])
	at com.fasterxml.jackson.databind.exc.InvalidFormatException.from(InvalidFormatException.java:67)
	at com.fasterxml.jackson.databind.DeserializationContext.weirdStringException(DeserializationContext.java:1549)
	at com.fasterxml.jackson.databind.DeserializationContext.handleWeirdStringValue(DeserializationContext.java:911)
	at com.fasterxml.jackson.datatype.jsr310.deser.JSR310DeserializerBase._handleDateTimeException(JSR310DeserializerBase.java:80)
	at com.fasterxml.jackson.datatype.jsr310.deser.InstantDeserializer.deserialize(InstantDeserializer.java:212)
	at com.fasterxml.jackson.datatype.jsr310.deser.InstantDeserializer.deserialize(InstantDeserializer.java:50)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:127)
	at io.fabric8.kubernetes.model.jackson.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:134)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:127)
	at io.fabric8.kubernetes.model.jackson.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:134)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:127)
	at io.fabric8.kubernetes.model.jackson.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:134)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:286)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:27)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:127)
	at io.fabric8.kubernetes.model.jackson.SettableBeanPropertyDelegate.deserializeAndSet(SettableBeanPropertyDelegate.java:134)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:1611)
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1188)
	at io.fabric8.kubernetes.client.utils.KubernetesSerialization.unmarshal(KubernetesSerialization.java:254)
	... 18 common frames omitted
	Caused by: java.time.format.DateTimeParseException: Text '2023-08-07T06:43:19+00:00' could not be parsed, unparsed text found at index 22
	at java.base/java.time.format.DateTimeFormatter.parseResolved0(DateTimeFormatter.java:2049)
	at java.base/java.time.format.DateTimeFormatter.parse(DateTimeFormatter.java:1874)
	at com.fasterxml.jackson.datatype.jsr310.deser.InstantDeserializer.deserialize(InstantDeserializer.java:206)
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by andreaTP**
## Description

Fix #5382 

Given the fact that the [specification](https://datatracker.ietf.org/doc/html/rfc3339#section-5.6) allows for even more special cases (e.g. using lower-case 't' and 's'), here we propose to cover the most common formats with defaults, but we enable the users to customize (separately) the format for serialization and deserialization of `date-time` fields.

## Type of change
 - [X] Bug fix (non-breaking change which fixes an issue)
 - [X] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [X] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/main/CHANGELOG.md) entry regarding this change
 - [X] I have implemented unit tests to cover my changes
 - [X] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/main/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

File: java-generator/maven-plugin/src/main/java/io/fabric8/java/generator/maven/plugin/JavaGeneratorMojo.java [hunk 1]
```diff
@@ -73,14 +73,6 @@ public class JavaGeneratorMojo extends AbstractMojo {
   @Parameter(property = "fabric8.java-generator.enum-uppercase", required = false)
   Boolean enumUppercase = null;
 
-  /**
```
Comment (andreaTP) [inline java-generator/maven-plugin/src/main/java/io/fabric8/java/generator/maven/plugin/JavaGeneratorMojo.java:76]: This is a leftover, not used around hence not a breaking change.

**Comment by andreaTP** [comment]
Ideally, this fix should be backported to a 6.8.1 release, cc. @manusa .

**Review by manusa — APPROVED** [review]

**Review by fabiobrz — APPROVED** [review]
HI @andreaTP - I've gone through the specification and checked the implementation changes, which LGTM. I dropped one very minor question about a potential formatting typo in the README, but approving anyway.

File: doc/java-generation-from-CRD.md [hunk 2]
```diff
@@ -80,12 +80,18 @@ The full list of options of the CLI is (output of `--help`):
 
 ```
 Usage: java-gen [-hV] [-add-extra-annotations] [-enum-uppercase]
-                [-dt=<downloadTarget>] [-s=<source>]
-                [-suffix-strategy=<suffixStrategy>] -t=<target>
+                [-deserialization-datetime-format=<deserializationDateTimeFormat
+                >] [-dt=<downloadTarget>] [-s=<source>]
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply doc/java-generation-from-CRD.md:84]: 🤷 that's what Maven prints out ... not really nice but shouldn't be an issue ...

**Review by rohanKanojia — APPROVED** [review]

**Comment by andreaTP** [comment]
For the records:

- the original formatting was working fine on Java 17+ but it was failing on Java 8 ( 🤯 )
- I improved the `DateTimeFormat` patterns and found a cross-compatible encoding
- I added more negative tests to more accurately verify that the output is expected

all in all, I'm happy to make this configurable so that users will be able to tweak it depending on their environment/use-case.

**Pull request merged**